src/MongoDB.Driver.Core/Core/ChannelPinningHelper.cs

```diff
@@ -0,0 +1,174 @@
+                }
+                readWriteBinding = new WritableServerBinding(cluster, session);
+            }
+
+            return new ReadWriteBindingHandle(readWriteBinding);
+        }
+
+        internal static IChannelSourceHandle CreateEffectiveGetMoreChannelSource(IChannelSourceHandle channelSource, long cursorId)
+        {
+            IChannelSource effectiveChannelSource;
+            if (IsInLoadBalancedMode(channelSource.ServerDescription) && cursorId != 0)
+            {
+                var getMoreChannel = channelSource.GetChannel(CancellationToken.None); // no need for cancellation token since we already have channel in the source
+                var getMoreSession = channelSource.Session.Fork();
+
+                effectiveChannelSource = new ChannelChannelSource(
+                    channelSource.Server,
+                    getMoreChannel,
+                    getMoreSession);
+            }
+            else
+            {
+                effectiveChannelSource = new ServerChannelSource(channelSource.Server, channelSource.Session.Fork());
+            }
+
+            return new ChannelSourceHandle(effectiveChannelSource);
+        }
+
+        internal static bool TryCreatePinnedChannelSourceAndPinChannel(
+            IChannelSourceHandle channelSource,
+            IChannelHandle channel,
+            ICoreSessionHandle session,
+            out (IChannelSourceHandle PinnedChannelSource, IChannelHandle Channel) pinnedChannel)
+        {
+            pinnedChannel = default;
+            if (IsInLoadBalancedMode(channel.ConnectionDescription))
+            {
+                var server = channelSource.Server;
+                var forkedChannel = channel.Fork();
+                var forkedSession = session.Fork();
+
+                var pinnedChannelSource = new ChannelSourceHandle(
+                    new ChannelChannelSource(
+                        server,
+                        forkedChannel, 
+                        forkedSession));
+
+                if (session.IsInTransaction)
+                {
+                    PinToTheSession(forkedChannel.Fork(), server, session);
+                }
+
+                pinnedChannel = (pinnedChannelSource, forkedChannel);
+
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Pin the channel and the server to the session if required.
+        /// </summary>
+        /// <param name="getChannelFunc">The protected channel getter.</param>
+        /// <param name="server">The server.</param>
+        /// <param name="session">The session.</param>
+        internal static void PinToTheSessionIfAlreadyNotUnpinned(Func<IChannelHandle> getChannelFunc, IServer server, ICoreSessionHandle session)
+        {
+            if (IsInLoadBalancedMode(server.Description) && IsConnectionPinned(session.CurrentTransaction))
```

Comment:
this method is called for commit and about transaction. The idea is if we already unpin the connection (because of failure) we should not pin it again. In the rest of other cases, it's emulating logic we have for all operation with retryable logic